applications/upgradeability-by-proxy.md

```diff
@@ -0,0 +1,141 @@
+# Open Grant Proposal
+
+* **Project Name:** Smart Contract Upgradeability
+* **Team Name:** Fractal?
+* **Payment Address:** lolcrypto
+
+*The above combination of your GitHub account submitting the application and payment address will be your unique identifier during the program. Please keep them safe.*
+
+^ wat
+
+## Project Overview :page_facing_up:
+
+### Overview
+
+Smart contract upgradeability is important, because bugs, and allowing for new
+features. Upgrades may be constrained, like needing to have votes to upgrade,
+or something. Very flexible, this proposal.
+
+It would be nice if it were easy to use this when building with the smart
+contracts pallet.
+
+Our team is interested in this because something.
+
+
+### Project Details
+
+Upgradeability can be implemented in user-land, or as part of the smart
+contracts pallet.
+
+We are opting to start this off by implementing it in user land to be able to
+iterate with more flexibility.
+
+When we're happy with the result, we'll see what can be migrated to ink as a
+directive to reduce boilerplate.
+
+We're going to start by implementing the proxy pattern. We need to figure out
+how to handle state (which will be stored in the proxy, probably?) migrations
+when the implementation is upgraded.
+
+We also need to figure out how to handle upgrade authorization. We want this to
+be customizable, so that each contract can have its own logic for permitting
+upgrades. Can the authorization code itself be upgradeable? I don't see why
+not. Should it be a third contract that is upgradeable by itself?
+
+There is the delegator pattern. We're going to start with a template that
+contains the migration/authorization code, but whose proxied methods need to be
+manually implemented by the smart contract developer. The idea is to focus on
+the mechanism itself and not on the syntactic sugar that ink should provide.
+Having this should allow folks to use it, even though it won't be super ergonomic.
+
+The second part will be attempting to extract this pattern into an ink directive,
+automatically generating the proxied methods. This brings us ergonomy of use, making
+upgradeability easy to opt in.
+
+Parts of this may be liftable to the smart contracts pallet. Out of scope
+though, for future plans.
+
+
+### Ecosystem Fit
+Are there any other projects similar to yours? If so, how is your project different?
```

Comment:
```suggestion
Since smart contract development is more accessible and less involved than runtime development, we expect most folks who build on Polkadot to benefit from this mechanism. Substrate offers a natural development path from smart contracts to runtime, so it's also likely that this will assist folks getting started with smart contracts even if they intend to move to a custom runtime in the future.

We didn't find related ideas in the https://github.com/w3f/General-Grants-Program or https://github.com/w3f/Open-Grants-Program repos. We also asked some folks in the community, none of whom were aware of similar projects.
```